physics/gcycle.F90

```diff
@@ -235,10 +241,16 @@ SUBROUTINE GCYCLE (nblks, Model, Grid, Sfcprop, Cldprop)
           Sfcprop(nb)%alvwf  (ix) = ALBFC1  (len + npts  )
           Sfcprop(nb)%alnsf  (ix) = ALBFC1  (len + npts*2)
           Sfcprop(nb)%alnwf  (ix) = ALBFC1  (len + npts*3)
-          do ls = 1,Model%lsoil
-            Sfcprop(nb)%smc (ix,ls) = SMCFC1 (len + (ls-1)*npts)
-            Sfcprop(nb)%stc (ix,ls) = STCFC1 (len + (ls-1)*npts)
-            Sfcprop(nb)%slc (ix,ls) = SLCFC1 (len + (ls-1)*npts)
+          do ls = 1,max(Model%lsoil,Model%lsoil_lsm)
+            if(Model%lsoil == Model%lsoil_lsm) then
+              Sfcprop(nb)%smc (ix,ls) = SMCFC1 (len + (ls-1)*npts)
+              Sfcprop(nb)%stc (ix,ls) = STCFC1 (len + (ls-1)*npts)
+              Sfcprop(nb)%slc (ix,ls) = SLCFC1 (len + (ls-1)*npts)
+            else
+              Sfcprop(nb)%smois (ix,ls) = SMCFC1 (len + (ls-1)*npts)
```

Comment:
Is this really generic? Can we assume that any land surface model that doesn't have `Model%lsoil == Model%lsoil_lsm` uses those RUC-specific arrays `smois`, `tslb`, `sh2o`?

Reply:
At this time we have two types of soil arrays: dimensioned 4 and dimentioned lsoil_lsm. Therefore I thought we could use the same names as in RUC LSM for lsoil_lsm. Do you have another option in mind?

Reply:
I am just wondering when we will make the switch and clean up the code so that every LSM uses the original arrays from Noah, and those being allocated correctly. Maybe the time is not now, though.

Reply:
we have to add the interpolation to initialize the soil arrays with flexible number of layers. You suggested it could be done in IO part, right? It would be great if we do this, but you think it could involve too many changes?

Reply:
> we have to add the interpolation to initialize the soil arrays with flexible number of layers. You suggested it could be done in IO part, right? It would be great if we do this, but you think it could involve too many changes?

I am wondering about the sequence of operations, too. Are the RUC soil layers initialized in the LSM RUC init calls or in the first time step? If in the first time step, then the gcycle.F90/sfcsub.F call for the first time step happens earlier and won't have any effect.

Reply:
Soil state interpolation is done at the first time step, the call is in sfc_drv_ruc.F90. We could not put it into RUC init, at least at the time when I worked on it.

Reply:
We should now be able to do this at init time. I wonder if it was better get rid of that entire duplication of soil arrays logic and do the interpolation when reading cold start initial conditions. We need to look at the code and decide how difficult it is.

Reply:
It sounds very attractive to have just one set of soil arrays. In this case the most part of rucinit that does vertical interpolation will not be needed.  The interpolation subroutine is inherited from WRF and is called  module_soil_pre.F90, which does all kind of interpolations between the parent and internal LSMs.